### PR TITLE
Report TestCoverage from Treadle and Verilator backends

### PR DESCRIPTION
**Pull request by ekiwi**
This PR changes the `test(...){...}` method to return a firrtl `AnnotationSeq`.
When using Treadle or Verilator as backends, this sequence will contain a `TestCoverage` annotation with coverage counts for every `cover` statement in the circuit.

This PR contains about 50% of the changes from https://github.com/ucb-bar/chisel-testers2/pull/274.
I tried to isolate the minimal changes necessary in `chiseltest` to support automated coverage, without adding any actual coverage passes just yet.

### CODE REVIEW

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorBackend.scala [hunk 1]
```diff
@@ -246,5 +251,15 @@ class VerilatorBackend[T <: Module](
 
       simApiInterface.finish() // Do this to close down the communication
     }
+    collectCoverageAnnotations()
```
Comment (chick) [inline src/main/scala/chiseltest/legacy/backends/verilator/VerilatorBackend.scala:254]: Nit: wondering if this should be `convertCoverageAnnotations` or similar to make clear it's distinction from other `collectCoverageAnnotations` methods

File: src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  val x = IO(Input(SInt(8.W)))
+  val y = IO(Input(SInt(8.W)))
+  val inCircle = IO(Output(Bool()))
+  val inRectangle = IO(Output(Bool()))
+  val radius = 100
+  inCircle := (x * x + y * y) <= (radius * radius).S
+  inRectangle := (x <= radius.S && x >= -radius.S) && (y <= radius.S && y >= -radius.S)
+  chisel3.experimental.verification.cover(inCircle)
+  chisel3.experimental.verification.cover(inRectangle)
+}
```
Comment (chick) [inline src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala:85]: missing newline

**Review by chick — APPROVED** [review]
Looks great to me. I have some comments in there, not any show stoppers. Probably good to get a cursory sign off from @ducky64

File: src/main/scala/chiseltest/coverage/ModuleInstancesPass.scala [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    val children = InstanceKeyGraph(state.circuit).getChildInstances.toMap
+    val topInstance = InstanceKey("", state.circuit.main)
+    val topChildren = children(topInstance.module)
+    val instances = topInstance +: topChildren.flatMap(onInstance("", _, children))
+    val instanceToModule = instances.toList.map(i => i.name -> i.module)
+    val anno = ModuleInstancesAnnotation(instanceToModule)
+    state.copy(annotations = anno +: state.annotations)
+  }
+
+  /** expands the instance name to its complete path (module the main module) */
```
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/chiseltest/coverage/ModuleInstancesPass.scala:42]: Good catch, I will try to fix this.

File: src/main/scala/chiseltest/internal/BackendInterface.scala [hunk 4]
```diff
@@ -118,7 +119,9 @@ trait BackendInstance[T <: Module] extends BackendInterface {
     * Takes the test function, which takes the module used as the testing interface.
     * TesterContext setup is done externally.
     *
+    * @return coverage annotations
```
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/chiseltest/internal/BackendInterface.scala:122]: Yes. The idea here is that we want to reduce the number of annotations that are kept around during the test in order to consume less main memory. Like if we just retained all annotations without any filtering, there could be a huge custom annotation that is kept alive and eats up all your RAM.

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorBackend.scala (hunk 1 above)
Comment (ekiwi) [reply src/main/scala/chiseltest/legacy/backends/verilator/VerilatorBackend.scala:254]: OK. I will rename it.

File: src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala (hunk 2 above)
Comment (ekiwi) [reply src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala:85]: I was confused for a bit why `scalafmt` didn't fix this. Turns out it does not run on the `test` directory, only on `main`.

File: src/main/scala/chiseltest/coverage/ModuleInstancesPass.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/chiseltest/coverage/ModuleInstancesPass.scala:42]: Turns out what I meant to say was _modul**o** the main module_

[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/chiseltest/coverage/ModuleInstancesPass.scala:42]: But I will try to rephrase it to be less confusing

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorBackend.scala (hunk 1 above)
Comment (ekiwi) [reply src/main/scala/chiseltest/legacy/backends/verilator/VerilatorBackend.scala]: fixed

File: src/main/scala/chiseltest/coverage/ModuleInstancesPass.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/chiseltest/coverage/ModuleInstancesPass.scala]: fixed

File: src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala (hunk 2 above)
Comment (ekiwi) [reply src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala]: fixed

File: src/main/scala/chiseltest/backends/treadle/TreadleBackend.scala [hunk 5]
```diff
@@ -14,7 +16,8 @@ class TreadleBackend[T <: Module](
   val dut:                T,
   val dataNames:          Map[Data, String],
   val combinationalPaths: Map[Data, Set[Data]],
-  tester:                 TreadleTester)
+  tester:                 TreadleTester,
+  coverageAnnotations:    AnnotationSeq)
```
Comment (ducky64) [inline src/main/scala/chiseltest/backends/treadle/TreadleBackend.scala:20]: Does it make sense for something more general here? (eg, pass in all annotations)

File: src/main/scala/chiseltest/coverage/Coverage.scala [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package chiseltest.coverage
+
+import firrtl._
+import firrtl.annotations.{Annotation, NoTargetAnnotation}
+
+/** Coverage counts returned from the simulator interface. */
+case class TestCoverage(counts: List[(String, Long)]) extends NoTargetAnnotation
```
Comment (ducky64) [inline src/main/scala/chiseltest/coverage/Coverage.scala:7]: Can you define (in a comment) what the strings mean?

File: src/test/scala/chiseltest/coverage/CompilerTest.scala [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import firrtl.options.TargetDirAnnotation
+import org.scalatest.TestSuite
+
+import java.io.File
+
+/** Base trait for tests that need to compile a circuit and inspect the resulting firrtl / Verilog  */
+trait CompilerTest extends ChiselScalatestTester { this: TestSuite =>
+  protected def annos: AnnotationSeq = Seq()
+
+  protected def compile[M <: Module](gen: => M, target: String, a: AnnotationSeq = List(), ll: String = "warn"): (String, AnnotationSeq) = {
```
Comment (ducky64) [inline src/test/scala/chiseltest/coverage/CompilerTest.scala:19]: Can you write a comment to define the return string-type? It appears it's any of FIRRTL or Verilog?

File: src/test/scala/chiseltest/coverage/Test1Module.scala [hunk 8]
```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (ducky64) [inline src/test/scala/chiseltest/coverage/Test1Module.scala:1]: It doesn't appear this is being used other than in SimulatorCoverageTest, can it be put in that same file?

File: src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    val cov = getCoverage(r)
+
+    // the cover point in the main module are not prefixed, but the one in the child module are
+    assert(cov.keys.toList.sorted == List("c0.cover_0", "c1.cover_0", "cover_0"))
+
+    // since we executed one step (+ the implicit reset) and all inputs are zero by default, we expect the count to be 3
+    assert(cov("cover_0") == 3)
+  }
+
+  it should "allow us to estimate pi" in {
```
Comment (ducky64) [inline src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala:47]: A short blurb about the intent of this test could be helpful. (what is this test testing? or is this just a catch-all integration test - still useful, but clarity is nice)

File: src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    val r = test(new Test1Module(withSubmodules = true)).withAnnotations(noAutoCov) { dut =>
+      dut.clock.step()
+    }
+    val cov = getCoverage(r)
+
+    // the cover point in the main module are not prefixed, but the one in the child module are
+    assert(cov.keys.toList.sorted == List("c0.cover_0", "c1.cover_0", "cover_0"))
+
+    // since we executed one step (+ the implicit reset) and all inputs are zero by default, we expect the count to be 3
+    assert(cov("cover_0") == 3)
```
Comment (ducky64) [inline src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala:44]: Do you also want to check the value for the child modules too?

File: src/test/scala/chiseltest/experimental/tests/VerilatorCoverageTests.scala [hunk 11]
```diff
@@ -23,9 +23,8 @@ class VerilatorCoverageTests extends AnyFlatSpec with ChiselScalatestTester with
     }
     val output = outputStream.toString
     coverage.exists() should be(true)
-    output.contains("--coverage-toggle") should be(true)
-    output.contains("--coverage-line") should be(false)
-    output.contains("--coverage-user") should be(false)
+    assert(output.contains("--coverage-toggle"))
```
Comment (ducky64) [inline src/test/scala/chiseltest/experimental/tests/VerilatorCoverageTests.scala:26]: Can you do something like `output should contain(...)`?
(and similarly below)

**Review by ducky64 — COMMENTED** [review]
Overall looks sane, with one concern about committing to an AnnotationSeq return type for test invocations. Also stylistic suggestions.

I didn't look over the code in the legacy directory.

File: src/main/scala/chiseltest/backends/treadle/TreadleBackend.scala (hunk 5 above)
Comment (ekiwi) [reply src/main/scala/chiseltest/backends/treadle/TreadleBackend.scala:20]: My assumption was that this is a semi-private API, i.e., most users aren't really expected to manually instantiate it. Thus changing this to be more broad in the future shouldn't be too big of a problem. For now this only contains the coverage annotations, so the name is very descriptive wrt. the current implementation.

File: src/main/scala/chiseltest/coverage/Coverage.scala (hunk 6 above)
Comment (ekiwi) [reply src/main/scala/chiseltest/coverage/Coverage.scala:7]: Will do.

File: src/test/scala/chiseltest/coverage/CompilerTest.scala (hunk 7 above)
Comment (ekiwi) [reply src/test/scala/chiseltest/coverage/CompilerTest.scala:19]: Yes.

File: src/test/scala/chiseltest/coverage/Test1Module.scala (hunk 8 above)
Comment (ekiwi) [reply src/test/scala/chiseltest/coverage/Test1Module.scala:1]: I would rather not, since once I add the line coverage, I will depend on exact line number in this file for my tests.
So yes, in isolation it would make sense to move this into its own file, but considering the other additions I want to make, it wouldn't be a good idea.

The one thing I can offer is to make a separate package, something along the lines of `chiseltest.coverage.circuits` or `chiseltest.coverage.examples` or `chiseltest.coverage.benchmarks` or similar.

File: src/test/scala/chiseltest/experimental/tests/VerilatorCoverageTests.scala (hunk 11 above)
Comment (ekiwi) [reply src/test/scala/chiseltest/experimental/tests/VerilatorCoverageTests.scala:26]: I am personally not a fan of the matchers and the original code was obviously using them wrong which meant that the error message I got from scalatest was not helpful at all. If you want me to, I can try to change it back to using matchers _correctly_, however I am not 100% certain that I know how to.

File: src/main/scala/chiseltest/ChiselScalatestTester.scala [hunk 12]
```diff
@@ -22,7 +22,7 @@ trait ChiselScalatestTester extends Assertions with TestSuiteMixin with TestEnvI
       sanitizeFileName(scalaTestContext.value.get.name)
     }
 
-    def apply(testFn: T => Unit): Unit = {
+    def apply(testFn: T => Unit): AnnotationSeq = {
```
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/chiseltest/ChiselScalatestTester.scala:25]: I was trying to make the minimal changes, so just directly returning an `AnnotationSeq` was the most straight forward thing to do. I can add a wraper object with a `.getAnnotationSeq` method if you would like me to.

File: src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala (hunk 10 above)
Comment (ekiwi) [reply src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala:44]: Good catch!

File: src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala (hunk 9 above)
Comment (ekiwi) [reply src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala:47]: added a comment

File: src/main/scala/chiseltest/backends/treadle/TreadleBackend.scala (hunk 5 above)
Comment (ducky64) [reply src/main/scala/chiseltest/backends/treadle/TreadleBackend.scala:20]: Yep, that's fine.

File: src/test/scala/chiseltest/coverage/Test1Module.scala (hunk 8 above)
Comment (ducky64) [reply src/test/scala/chiseltest/coverage/Test1Module.scala:1]: I see. The main benefit of having it in the same file would be marginally more convenient readability. But a comment describing the line stability concerns would be great.

File: src/test/scala/chiseltest/experimental/tests/VerilatorCoverageTests.scala (hunk 11 above)
Comment (ducky64) [reply src/test/scala/chiseltest/experimental/tests/VerilatorCoverageTests.scala:26]: No strong preference as long as we're somewhat locally consistent.

File: src/main/scala/chiseltest/ChiselScalatestTester.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (ducky64) [reply src/main/scala/chiseltest/ChiselScalatestTester.scala:25]: I think having a wrapper class would be preferred to keep flexibility.

File: src/test/scala/chiseltest/experimental/tests/VerilatorCoverageTests.scala (hunk 11 above)
Comment (ekiwi) [reply src/test/scala/chiseltest/experimental/tests/VerilatorCoverageTests.scala:26]: I converted it to matchers.
The right matcher is `output should include(...)` since `contain` only seems to work on sequences, but not on a String.

File: src/main/scala/chiseltest/ChiselScalatestTester.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/chiseltest/ChiselScalatestTester.scala:25]: I am creating a `TestResults` wrapper class. Do you have a good idea where it should live?
I was going to put it into the `Testers2.scala`, but then I realized that it is in the `internal` package and
we would probably want `TestResults to just live in `chiseltest.`.

File: src/test/scala/chiseltest/coverage/Test1Module.scala (hunk 8 above)
Comment (ekiwi) [reply src/test/scala/chiseltest/coverage/Test1Module.scala:1]: added a comment

File: src/main/scala/chiseltest/TestResult.scala [hunk 13]
```diff
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chiseltest
+
+import firrtl.AnnotationSeq
+
+class TestResult(annotations: AnnotationSeq) {
```
Comment (ekiwi) [inline src/main/scala/chiseltest/TestResult.scala:7]: @ducky64 Is there a good file to put the `TestResult` class in? (Instead of creating a new file)

Comment (ducky64) [reply src/main/scala/chiseltest/TestResult.scala:7]: I think having its own file is fine? It is (or I guess could be) a significant API, even though the implementation is super short?

Comment (ekiwi) [reply src/main/scala/chiseltest/TestResult.scala:7]: I am happy with that.
Any good ideas for a non-obvious scala doc?
I was trying, but all I could come up with was:
> Contains the results of a test run.

Comment (ducky64) [reply src/main/scala/chiseltest/TestResult.scala:7]: Perhaps also mention that this is meant for future expansion of the test return API, though there's not much in it right now.

Comment (ducky64) [reply src/main/scala/chiseltest/TestResult.scala:7]: (I could totally see someone asking "wait why does this even exist")

Comment (ekiwi) [reply src/main/scala/chiseltest/TestResult.scala:7]: I will add a plain comment, since this is mostly developer and not user facing.

Comment (ekiwi) [reply src/main/scala/chiseltest/TestResult.scala:9]: added the comment

**Review by ducky64 — APPROVED** [review]
lgtm!
